call-home/src/bin/callhome/collector/report_models.rs

```diff
@@ -184,3 +199,160 @@ fn get_pools_size_vector(pools: Vec<openapi::models::Pool>) -> Vec<u64> {
     }
     pools_size_vector
 }
+
+/// EventData consists of pool and volume events.
+#[derive(Debug, Default, Clone)]
+pub struct EventData {
+    pub volume_created: VolumeCreated,
+    pub volume_deleted: VolumeDeleted,
+    pub pool_created: PoolCreated,
+    pub pool_deleted: PoolDeleted,
+}
+
+/// Number of volume created event.
+#[derive(Debug, Default, Clone)]
+pub struct VolumeCreated(u32);
+
+impl VolumeCreated {
+    // Method to retrieve the value
+    fn value(&self) -> u32 {
+        self.0
+    }
+}
+
+/// Number of volume deleted event.
+#[derive(Debug, Default, Clone)]
+pub struct VolumeDeleted(u32);
+
+impl VolumeDeleted {
+    // Method to retrieve the value
+    fn value(&self) -> u32 {
+        self.0
+    }
+}
+
+/// Number of pool created event.
+#[derive(Debug, Default, Clone)]
+pub struct PoolCreated(u32);
+
+impl PoolCreated {
+    // Method to retrieve the value
+    fn value(&self) -> u32 {
+        self.0
+    }
+}
+
+/// Number of pool deleted event.
+#[derive(Debug, Default, Clone)]
+pub struct PoolDeleted(u32);
+
+impl PoolDeleted {
+    // Method to retrieve the value
+    fn value(&self) -> u32 {
+        self.0
+    }
+}
+
+/// Record of events populated from prometheus.
+#[derive(Debug, Default, Clone)]
+/// Wrapper for Stats fetched from prometheus data.
+pub struct EventsRecord {
+    pub record: Vec<Sample>,
+}
+
+/// Implements methods for wrapper.
+impl EventsRecord {
+    pub fn new(data: Vec<Sample>) -> Self {
+        Self { record: data }
+    }
+
+    /// Returns VolumeCreated.
+    pub fn volume_created(&self) -> VolumeCreated {
+        let mut vc = 0;
+        for rec in self.record.iter() {
+            let value = rec.labels.get(ACTION).unwrap_or_default();
+            if value == CREATED && rec.metric.as_str() == VOLUME {
+                vc = stats(rec.value.clone());
+                break;
```

Comment:
hmm don't think you can break here, what if other records are also volume?
I suggest something like this:
```rust
struct Record(extern_type);
impl Record {
   fn action(&self) -> ACTION {
     self.labels.get(ACTION).unwrap_or_default()
   }
   fn metric(&self) -> Option<stats::MetricType> {
     self.metric.as_str().try_into().ok()
   }
   fn gauge_value(&self) -> u32 { 
     match val {
        Value::Gauge(v) => v as u32,
        _ => 0,
     }
   }
}
impl TryFrom<Record> for VolumeCreated {
   type Error=();
   fn try_from(src: &Record) -> Result<Self, ()> {
      match src.metric() {
       Some(stats::Volume) => {
         Ok(VolumeCreated(src.gauge_value())
       }
       _ => None,
      }
   }
}
impl<T: TryFrom<Record>> From<Vec<Record>> for Vec<T> {
  fn from(src: &Vec<Record>) -> Vec<VolumeCreated> {
    self.iter().filter_map(|r| r.try_from().ok()).collect::Vec<_>()
  }
}

fn data_thingy() {
let volumes_created: Vec<VolumesCreated> = records.into();
let pools_created: Vec<PoolsCreated> = records.into();
}
```

Reply:
Thanks for the help @tiagolobocastro . addressed your comment